File: valid-parentheses.js
// https://leetcode.com/problems/valid-parentheses/
// https://leetcode.com/submissions/detail/271044431/

/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    if(s === undefined || typeof s != "string") throw new Error("Invalid Input");

    s = s.split('').reverse();

    let stack = [];

    for(let i = 0; i < s.length; i++){
        stack.push(s[i]);

        if(i>0){
            if( ( stack[stack.length-1] == '(' && stack[stack.length-2] == ')' ) ||
                ( stack[stack.length-1] == '[' && stack[stack.length-2] == ']' ) ||
                ( stack[stack.length-1] == '{' && stack[stack.length-2] == '}' ) ) {
                stack.pop();
                stack.pop();
            }
        }
    }

    return stack.length == 0;
};

File: python/Stack/valid-parentheses.py
# https://leetcode.com/problems/valid-parentheses/submissions/

from collections import deque 

class Solution:
    def isValid(self, s: str) -> bool:
        
        dictionary = {
            '(' : ')',
            '{' : '}',
            '[' : ']'
        }
        
        stack = deque()
        
        for char in s:
            if char in dictionary:
                stack.append(char)
            elif stack and stack[-1] in dictionary and char == dictionary[stack[-1]]:
                stack.pop()
            else:
                stack.append(char)

        return True if len(stack) == 0 else False